File: os-in-rust/src/keyboard.rs
pub const ALPHABET: [u8; NUM_CHARS] = [
    b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', b'i', b'j', b'k', b'l', b'm', b'n', b'o', b'p',
    b'q', b'r', b's', b't', b'u', b'v', b'w', b'x', b'y', b'z', b'1', b'2', b'3', b'4', b'5', b'6',
    b'7', b'8', b'9', b'0', b'.', b'\n', b' ',
];

pub const NUM_CHARS: usize = 39;
pub const BACKSPACE_BIT: usize = 39;

pub struct Keyboard {
    keys_down: u64,
}

impl Keyboard {
    pub fn new() -> Self {
        Self { keys_down: 0 }
    }

    pub fn update(&mut self) {
        let mut x: u8 = 0;
        unsafe {
            asm!(
                "in al, 0x60",
                "mov {}, al",
                inout(reg_byte) x,
            );
        }
        let action = Action::from_qwerty_u8(x);
        let mut set_bit = |enable: bool, n: usize| {
            if enable {
                self.keys_down |= 1 << n;
            } else {
                self.keys_down &= !(1 << n);
            }
        };

        match action {
            Action::Unrecognized => {}

            Action::A(bool) => set_bit(bool, 0),
            Action::B(bool) => set_bit(bool, 1),
            Action::C(bool) => set_bit(bool, 2),
            Action::D(bool) => set_bit(bool, 3),
            Action::E(bool) => set_bit(bool, 4),
            Action::F(bool) => set_bit(bool, 5),
            Action::G(bool) => set_bit(bool, 6),
            Action::H(bool) => set_bit(bool, 7),
            Action::I(bool) => set_bit(bool, 8),
            Action::J(bool) => set_bit(bool, 9),
            Action::K(bool) => set_bit(bool, 10),
            Action::L(bool) => set_bit(bool, 11),
            Action::M(bool) => set_bit(bool, 12),
            Action::N(bool) => set_bit(bool, 13),
            Action::O(bool) => set_bit(bool, 14),
            Action::P(bool) => set_bit(bool, 15),
            Action::Q(bool) => set_bit(bool, 16),
            Action::R(bool) => set_bit(bool, 17),
            Action::S(bool) => set_bit(bool, 18),
            Action::T(bool) => set_bit(bool, 19),
            Action::U(bool) => set_bit(bool, 20),
            Action::V(bool) => set_bit(bool, 21),
            Action::W(bool) => set_bit(bool, 22),
            Action::X(bool) => set_bit(bool, 23),
            Action::Y(bool) => set_bit(bool, 24),
            Action::Z(bool) => set_bit(bool, 25),

            Action::One(bool) => set_bit(bool, 26),
            Action::Two(bool) => set_bit(bool, 27),
            Action::Three(bool) => set_bit(bool, 28),
            Action::Four(bool) => set_bit(bool, 29),
            Action::Five(bool) => set_bit(bool, 30),
            Action::Six(bool) => set_bit(bool, 31),
            Action::Seven(bool) => set_bit(bool, 32),
            Action::Eight(bool) => set_bit(bool, 33),
            Action::Nine(bool) => set_bit(bool, 34),
            Action::Zero(bool) => set_bit(bool, 35),
            Action::Point(bool) => set_bit(bool, 36),

            Action::Enter(bool) => set_bit(bool, 37),
            Action::Space(bool) => set_bit(bool, 38),

            Action::Backspace(bool) => set_bit(bool, 39),
        }
    }

    pub fn state(&self) -> u64 {
        self.keys_down
    }
}

enum Action {
    Unrecognized,

    Q(bool),
    W(bool),
    E(bool),
    R(bool),
    T(bool),
    Y(bool),
    U(bool),
    I(bool),
    O(bool),
    P(bool),
    A(bool),
    S(bool),
    D(bool),
    F(bool),
    G(bool),
    H(bool),
    J(bool),
    K(bool),
    L(bool),
    Z(bool),
    X(bool),
    C(bool),
    V(bool),
    B(bool),
    N(bool),
    M(bool),

    One(bool),
    Two(bool),
    Three(bool),
    Four(bool),
    Five(bool),
    Six(bool),
    Seven(bool),
    Eight(bool),
    Nine(bool),
    Zero(bool),

    Point(bool),

    Backspace(bool),
    Space(bool),
    Enter(bool),
}

impl Action {
    pub fn from_qwerty_u8(byte: u8) -> Self {
        let is_pressed = byte & 1 << 7 != 0;
        let byte = byte & !(1 << 7);
        match byte {
            0x10 => Action::Q(is_pressed),
            0x11 => Action::W(is_pressed),
            0x12 => Action::E(is_pressed),
            0x13 => Action::R(is_pressed),
            0x14 => Action::T(is_pressed),
            0x15 => Action::Y(is_pressed),
            0x16 => Action::U(is_pressed),
            0x17 => Action::I(is_pressed),
            0x18 => Action::O(is_pressed),
            0x19 => Action::P(is_pressed),
            0x1E => Action::A(is_pressed),
            0x1F => Action::S(is_pressed),
            0x20 => Action::D(is_pressed),
            0x21 => Action::F(is_pressed),
            0x22 => Action::G(is_pressed),
            0x23 => Action::H(is_pressed),
            0x24 => Action::J(is_pressed),
            0x25 => Action::K(is_pressed),
            0x26 => Action::L(is_pressed),
            0x2C => Action::Z(is_pressed),
            0x2D => Action::X(is_pressed),
            0x2E => Action::C(is_pressed),
            0x2F => Action::V(is_pressed),
            0x30 => Action::B(is_pressed),
            0x31 => Action::N(is_pressed),
            0x32 => Action::M(is_pressed),

            0x2 => Action::One(is_pressed),

            0x3 => Action::Two(is_pressed),

            0x4 => Action::Three(is_pressed),

            0x5 => Action::Four(is_pressed),

            0x6 => Action::Five(is_pressed),

            0x7 => Action::Six(is_pressed),

            0x8 => Action::Seven(is_pressed),

            0x9 => Action::Eight(is_pressed),

            0xA => Action::Nine(is_pressed),

            0xB => Action::Zero(is_pressed),

            0x34 => Action::Point(is_pressed),

            0xE => Action::Backspace(is_pressed),
            0x39 => Action::Space(is_pressed),
            0x1C => Action::Enter(is_pressed),

            _ => Action::Unrecognized,
        }
    }
}

File: bingbong/kernel/keyboard.c
#include "keyboard.h"
#include "common.h"
#include "vga.h"

static const u16 keyboard_data_port = 0x60;
static const u16 keyboard_status_port = 0x64;

u8 port_read(u16 port)
{
    u8 result;
    __asm__("mov dx, %[port]\t\n"
            "in %[result], dx\t\n"
            : [result] "=r"(result)
            : [port] "r"(port));
    return result;
}

void port_write(u16 port, u8 value)
{
    __asm__("mov dx, %[port]\t\n"
            "mov al, %[value]\t\n"
            "out dx, al\t\n"
            :
            : [port] "r"(port), [value] "r"(value));
}

void load_idt(void* address)
{
    __asm__("lidt [%[address]]\t\n"
            "sti\t\n"
            :
            : [address] "r"(address));
}

char qwerty_key_char_value(u8 key)
{
    u8 key_char = key & ~(1 << 7);
    // clang-format off
    switch (key_char) {
        case 0x2: return '1';  case 0x3: return '2';  case 0x4: return '3';
        case 0x5: return '4';  case 0x6: return '5';  case 0x7: return '6';
        case 0x8: return '7';  case 0x9: return '8';  case 0xa: return '9';
        case 0xb: return '0';  case 0xe: return '\b'; case 0x1c: return '\n';
        case 0x10: return 'q'; case 0x11: return 'w'; case 0x12: return 'e';
        case 0x13: return 'r'; case 0x14: return 't'; case 0x15: return 'y';
        case 0x16: return 'u'; case 0x17: return 'i'; case 0x18: return 'o';
        case 0x19: return 'p'; case 0x1e: return 'a'; case 0x1f: return 's';
        case 0x20: return 'd'; case 0x21: return 'f'; case 0x22: return 'g';
        case 0x23: return 'h'; case 0x24: return 'j'; case 0x25: return 'k';
        case 0x26: return 'l'; case 0x2c: return 'z'; case 0x2d: return 'x';
        case 0x2e: return 'c'; case 0x2f: return 'v'; case 0x30: return 'b';
        case 0x31: return 'n'; case 0x32: return 'm'; case 0x34: return '.';
        case 0x39: return ' '; default: return '\0';
    }
    // clang-format on
}

bool qwerty_key_is_pressed(u8 key) { return (key >> 7) == 0; }

void keyboard_create(Keyboard* self)
{
    *self = (Keyboard) {
        .keys_pressed = { 0 },
    };
}

void keyboard_update(Keyboard* self)
{
    u8 key = port_read(keyboard_data_port);
    char char_value = qwerty_key_char_value(key);
    bool is_pressed = qwerty_key_is_pressed(key);
    if (is_pressed) {
        if (char_value < 64) {
            self->keys_pressed[0] |= 1 << char_value;
        } else {
            self->keys_pressed[1] |= 1 << (char_value - 64);
        }
    } else {
        if (char_value < 64) {
            self->keys_pressed[0] &= ~(1 << char_value);
        } else {
            self->keys_pressed[1] &= ~(1 << char_value);
        }
    }
}

bool keyboard_key_was_pressed(Keyboard* self) { return false; }

char keyboard_pressed_char(Keyboard* self) { return '\0'; }

bool keyboard_key_was_released(Keyboard* self) { return false; }

char keyboard_released_char(Keyboard* self) { return '\0'; }
